crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -0,0 +1,1714 @@
+            instance_limits,
+            instances: mem::ManuallyDrop::new(instances),
+            stacks: mem::ManuallyDrop::new(stacks),
+            #[cfg(all(feature = "uffd", target_os = "linux"))]
+            _fault_handler,
+        })
+    }
+}
+
+impl Drop for PoolingInstanceAllocator {
+    fn drop(&mut self) {
+        // Manually drop the pools before the fault handler (if uffd is enabled)
+        // This ensures that any fault handler thread monitoring the pool memory terminates
+        unsafe {
+            mem::ManuallyDrop::drop(&mut self.instances);
+            mem::ManuallyDrop::drop(&mut self.stacks);
+        }
+    }
+}
+
+unsafe impl InstanceAllocator for PoolingInstanceAllocator {
+    fn validate_module(&self, translation: &ModuleTranslation) -> Result<(), String> {
+        self.module_limits.validate_module(&translation.module)
+    }
+
+    fn adjust_tunables(&self, tunables: &mut Tunables) {
+        let memory_reservation_size = self.instance_limits.memory_reservation_size;
+
+        // For reservation sizes larger than 4 GiB, use a guard region to elide bounds checks
+        if memory_reservation_size >= 0x100000000 {
+            tunables.static_memory_bound = 0x10000; // in Wasm pages
+            tunables.static_memory_offset_guard_size = memory_reservation_size - 0x100000000;
+        } else {
+            tunables.static_memory_bound =
+                u32::try_from(memory_reservation_size).unwrap() / WASM_PAGE_SIZE;
+            tunables.static_memory_offset_guard_size = 0;
+        }
+
+        // Treat the static memory bound as the maximum for unbounded Wasm memories
+        // Because we guarantee a module cannot compile unless it fits in the limits of
+        // the pool allocator, this ensures all memories are treated as static (i.e. immovable).
+        tunables.static_memory_bound_is_maximum = true;
+    }
+
+    unsafe fn allocate(
+        &self,
+        req: InstanceAllocationRequest,
+    ) -> Result<InstanceHandle, InstantiationError> {
+        self.instances.allocate(self.strategy, req)
+    }
+
+    unsafe fn initialize(
+        &self,
+        handle: &InstanceHandle,
+        is_bulk_memory: bool,
+    ) -> Result<(), InstantiationError> {
+        let instance = handle.instance();
+
+        cfg_if::cfg_if! {
+            if #[cfg(all(feature = "uffd", target_os = "linux"))] {
+                match instance.module.memory_initialization {
+                    Some(MemoryInitialization::Paged{ .. }) => {
+                        if !is_bulk_memory {
+                            check_init_bounds(instance)?;
+                        }
+
+                        // Initialize the tables
+                        initialize_tables(instance)?;
+
+                        // Don't initialize the memory; the fault handler will fill the pages when accessed
+                        Ok(())
+                    },
+                    _ => initialize_instance(instance, is_bulk_memory)
```

Comment:
If I understand this correctly, when uffd is enabled, this I think has somewhat weirder performance? During this standard initialization what will happen is that we'll fault as we write in each byte, but the uffd handler fills in zeros, and then we proceed to the next page. 

For this sort of initialization would it be better to go ahead and make the pages backed by real memory before we initialize the memories of the instance?

Reply:
For some context, here's my understanding of how linear memories are created and initialized in Wasmtime (on Linux) with these changes:

* On-demand allocator: by default we create a 6 GiB anonymous, private mapping with a `none` protection for the entire range.  We then mark the pages that are part of the "minimal" initial set for the linear memory as `read+write`.  We then iterate the data segments and they instruct us to write to one of those initial pages; but the page is not yet backed, so this causes a minor page fault and the kernel's memory manager handles it by backing the page with a single, shared CoW-marked zero-filled page. The write causes a copy of the page and thus a new page is physically backed for the process and subsequently filled with segment's data.  When the instance is deallocated, the entire mapping of the linear memory is freed.

* Pooling allocator: we allocate the same 6 GiB anonymous, private mapping with a `none` protection, but upfront and part of a pool of linear memories.  This is otherwise identical to the on-demand case: we still change the protection when creating/growing the memory and expect the kernel to handle missing backings with a zeroed page. Unlike the on-demand allocator, we may reuse the linear memory's mapping for a new instance; we effectively clear any backing for the linear memory's pages with `MADV_DONTNEED` when the instance is freed, thus guaranteeing the same initially-zero page the next time it is accessed (causing another minor fault).

* Pooling allocator with `uffd` enabled: same as without the `uffd` feature, except the entire region is protected as `read+write` instead of `none`.  This means we don't have to call `mprotect` to grow memories or when we reuse the linear memory's space for a new instance. Instead, we instruct the kernel to pass us any faults in this region and if we determine the access is out of bounds, we'll change the protection (temporarily) to `none` as-needed to induce a `SIGSEGV`.  We have two ways to handle the fault to an accessible page: zero the page or copy from another page.  As we're able to handle the faults ourselves, we can skip any initialization of defined linear memories at instantiation time and just do it when the page is first accessed. 

The trade-off for uffd is that handling page faults in user space is slower than letting the kernel handle the faults, but we can skip any data segment initialization until actually accessed (a win if there are large data segments that aren't fully utilized). But more importantly, without all those `mprotect` calls, we're reducing kernel lock contention, page table modifications, and (possibly?) TLB shootdowns.  Access to guard pages notwithstanding, the page table entries only get modified when an instance is freed with the `uffd` feature.

In theory, the penalty of handling the faults in user space is outweighed by the improved concurrency.

Reply:
Ok that all makes sense, thanks for explaining!

So basically uffd or not, the same thing happens in this case where we manually initialize all data segments, which is that the first fault on each page *someone* puts a page of zeros there and then we overwrite it.

My pondering here was mostly around how we know how big our memcpy is and which pages need to be resident, so I figured we could skip the uffd handler entirely and just tell the kernel "put a bunch of pages there", but that's probably overkill and not actually really needed for any situation today in terms of being a bottleneck.

Reply:
With wizer (I believe) organizing initialized memories on page boundaries, I think an implementation that mmaps these pages CoW directly from the file might definitely be something to investigate for the future.  We could do the same for our cached compiled modules too.

Granted the whole bulk memory side-effects thing might throw this stuff for a loop).

Reply:
> With wizer (I believe) organizing initialized memories on page boundaries

Correct.

It is also perhaps worth doing this kind of pre-processing of all modules' data segments when we compile them, so that we could hit this optimization unconditionally.

Reply:
> Granted the whole bulk memory side-effects thing might throw this stuff for a loop.

Should this turn out to really throw a wrench in the works here, I wonder if that might be a reason to revisit this at the spec level? I know that bulk memory is stable, but this seems like a corner-case-y enough thing that perhaps there's some wiggle room?

Reply:
I think we can still do the preprocessing, similar to what this PR is doing in-memory, but due to the bulk memory spec's relaxation of the upfront bounds checking at initialization time, we'll have to enforce certain requirements to initialize memory this way:

* If any data segment references an imported table, use the existing memory initialization implementation (i.e. iterate the segments directly).
* If any data segment uses a global base, use the existing implementation (global bases, while immutable, are always an import).
* If neither of those are true, then we can organize the data segments into pages at compilation time; if a data segment is out of bounds, we stop the organization and set a flag indicating a segment went out of bounds.

When initializing the instance, we would initialize tables first per the spec; an out of bounds element segment would fail the instantiation and memories wouldn't be initialized at all.

When initializing the memories, if we're using the existing initialization strategy, we iterate the segments and copy the data, stopping if the segment was out of bounds and thus failing the instantiation.

If we're using "paged" initialization, we copy (delayed for uffd) or, potentially, mmap all of the pages in the organized data and then check the flag to see if there was an out of bounds segment, returning failure to instantiate if that's the case.  This should maintain the "partially initialized" side effects allowed for in the bulk memory proposal, I think.